core/state/journal_linear.go

```diff
@@ -0,0 +1,500 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"maps"
+	"slices"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/holiman/uint256"
+)
+
+// journalEntry is a modification entry in the state change linear journal that can be
+// reverted on demand.
+type journalEntry interface {
+	// revert undoes the changes introduced by this entry.
+	revert(*StateDB)
+
+	// dirtied returns the Ethereum address modified by this entry.
+	dirtied() *common.Address
+
+	// copy returns a deep-copied entry.
+	copy() journalEntry
+}
+
+// linearJournal contains the list of state modifications applied since the last state
+// commit. These are tracked to be able to be reverted in the case of an execution
+// exception or request for reversal.
+type linearJournal struct {
+	entries []journalEntry         // Current changes tracked by the linearJournal
+	dirties map[common.Address]int // Dirty accounts and the number of changes
+
+	revisions []int // sequence of indexes to points in time designating snapshots
+}
+
+// compile-time interface check
+var _ journal = (*linearJournal)(nil)
+
+// newLinearJournal creates a new initialized linearJournal.
+func newLinearJournal() *linearJournal {
+	return &linearJournal{
+		dirties: make(map[common.Address]int),
+	}
+}
+
+// reset clears the journal, after this operation the journal can be used anew.
+// It is semantically similar to calling 'newJournal', but the underlying slices
+// can be reused.
+func (j *linearJournal) reset() {
+	j.entries = j.entries[:0]
+	j.revisions = j.revisions[:0]
+	clear(j.dirties)
+}
+
+func (j linearJournal) dirtyAccounts() []common.Address {
+	dirty := make([]common.Address, 0, len(j.dirties))
+	// flatten into list
+	for addr := range j.dirties {
+		dirty = append(dirty, addr)
+	}
+	return dirty
+}
+
+// snapshot starts a new journal scope which can be reverted or discarded.
+func (j *linearJournal) snapshot() {
+	j.revisions = append(j.revisions, len(j.entries))
+}
+
+// revertSnapshot reverts all state changes made since the last call to snapshot().
+func (j *linearJournal) revertSnapshot(s *StateDB) {
+	id := len(j.revisions) - 1
```

Comment:
Same here, this will panic if there was no previous call to snapshot()

Reply:
Interestingly enough the set based journaling will not panic on revertSnapshot or discardSnapshot